### PR TITLE
Cache table formatting funcs in col.info not pprint module

### LINKED ISSUE #6703: Table column print format cache bug
The column format is currently cached in a global dictionary keyed with the format and `id()` of the column.  The problem is that when the table/column get garbage collected, a totally different column can be later placed in the same memory location and thus have the same `id()`.  That may seem unlikely, but I've been encountering this seemingly random bug in real code (and in `photutils` CI testing).

The specific issue that I'm having is that a mixin `Quantity` column caches the formatter to use the one for `Quantity` objects: https://github.com/astropy/astropy/blob/master/astropy/units/quantity.py#L121

Later, a regular non-mixin `Column` cannot be printed because it's trying to use the `Quantity` formatter.  This is because the initial column has been garbage collected and the later `Column` has the same `id()` -- thus it uses the cached formatter.

This was frustrating to debug in actual code because it was hard to reliably reproduce, but here's a contrived example showing the bug:

https://gist.github.com/larrybradley/57185c001ae12647d74d9a73a53f6a5e

My suggestion would be to cache the formatter in the column `info` object, as suggested in #5804.  I don't understand the concern about the case when the column does not yet have data. Perhaps I missed something, but it appears that the caching occurs only when the first item of the column is printed, so the column must have data in order for the format to be cached.  Also, how does one even set the column format without having defined the column object?

Also related to:  #5800 and #5803

CC: @taldcroft, @mhvk

### PR DESCRIPTION
**Pull request by taldcroft**
Fixes #6703 

"The column format is currently cached in a global dictionary keyed with the format and id() of the column. The problem is that when the table/column get garbage collected, a totally different column can be later placed in the same memory location and thus have the same id(). That may seem unlikely, but I've been encountering this seemingly random bug in real code (and in photutils CI testing)."

This PR moves the cache dict (`_format_funcs`) into the column `Info` object instead of being a common dict in the `pprint` module.  This should fix the issue since the cache is specific to the column.  I added a test but was not able to reproduce the problem on my Mac with current master.  It would be good if @larrybradley can try demonstrating that the new test fails on master but passes with this PR.

The 7369e5a commit fixes a minor unrelated bug where the exception catching for a bad format actually raised an unexpected exception for a mixin column that has no `name` attribute.  This is a leftover of adding mixing columns where one should always use `col.info.name` not `col.name`.

### CODE REVIEW

**Comment by taldcroft** [comment]
This does make an API change to `get_auto_format_func`, but this is not formally part of the public API and does not appear in the API docs.  So I think it should be OK.

**Comment by taldcroft** [comment]
@mhvk - if @larrybradley cannot get this test to fail in master then I would agree to cut it out.  Or maybe just cut anyway if his demonstration jupyter notebook fails with master but works with this PR.  I am getting more aware these days of test times and ensuring that they are meaningful.

**Comment by larrybradley** [comment]
Thanks, @taldcroft!  I cannot get the new test to fail for this PR or master.  However, my notebook fails for master, but passes with this PR.  Feel free to remove the test since this bug occurs randomly.

File: astropy/utils/data_info.py [hunk 1]
```diff
@@ -446,6 +446,14 @@ class BaseColumnInfo(DataInfo):
     # depending on context.
     _serialize_context = None
 
+    def __init__(self, bound=False):
+        super().__init__(bound=bound)
```
Comment (MSeifert04) [inline astropy/utils/data_info.py]: If this is backported it needs the arguments for `super`. Regard this as just a note for the "backporter".

Comment (bsipocz) [reply astropy/utils/data_info.py]: @MSeifert04 - Please use the "manual backport" label for these cases, as I'm not reading through the comments when doing backports ;)

Comment (taldcroft) [reply astropy/utils/data_info.py]: And so begins the predicted pain of backporting Py3-only code to the LTS...  hopefully it will be manageable.

**Comment by taldcroft** [comment]
Note, since the last force push only deleted a new test (that was deemed useless), I stopped Travis after the initial tests.

File: astropy/utils/data_info.py (hunk 1 above)
Comment (bsipocz) [reply astropy/utils/data_info.py]: Funnily enough, this was quite painless, and the real conflicts are in pprint.py...

(Though the heads up is really needed as this part obviously wasn't conflicting)

**Pull request merged**